pennylane/transforms/hamiltonian_expand.py

```diff
@@ -44,13 +44,14 @@ def grouping_processing_fn(res_groupings, coeff_groupings, batch_size, offset):
             r_group = qml.math.stack(r_group)
         if qml.math.shape(r_group) == ():
             r_group = qml.math.reshape(r_group, (1,))
-        if batch_size:
-            r_group = r_group.T
+        if batch_size and batch_size > 1 and len(c_group) > 1:
+            r_group = qml.math.moveaxis(r_group, -1, -2)
```

Comment:
Basically the dimension of `r_group` here is `(group_size, batch_size)`, so taking the transpose was needed to successfully performing the dot product between `r_group` and `c_group`, which has dimension `(group_size,)`. However, this does not take into account shot batching. If shot vectors are involved, `r_group` would have the shape `(shot_vector_size, group_size, batch_size)`. In this case, `r_group.T` gives `(batch_size, group_size, shot_vector_size)`, which does not work. What we want to do is only transposing the last two dimensions such that the `group_size` is the last dimension where the dot product is applied, without changing the ordering of the rest of the batching dimensions.